File: packages/runtime-web-three/src/systems/componentDiff.ts
import type { IWorldEntity, IWorldIr } from "@threenative/ir";

export interface IComponentDiffCache {
  beginScheduleStage(world: IWorldIr, components: ReadonlyArray<string>): void;
  runtimeChangedComponents(entity: IWorldEntity): string[];
}

export function createComponentDiffCache(): IComponentDiffCache {
  let baseline = new Map<string, Map<string, string>>();
  let tracked = new Set<string>();

  return {
    beginScheduleStage(world, components) {
      tracked = new Set(components);
      baseline = new Map();
      for (const entity of world.entities) {
        for (const component of tracked) {
          const value = entity.components[component];
          if (value === undefined) {
            continue;
          }
          const row = baseline.get(entity.id) ?? new Map<string, string>();
          row.set(component, JSON.stringify(value));
          baseline.set(entity.id, row);
        }
      }
    },
    runtimeChangedComponents(entity) {
      const changed: string[] = [];
      for (const component of [...tracked].sort()) {
        const current = entity.components[component];
        const currentValue = current === undefined ? undefined : JSON.stringify(current);
        const baselineValue = baseline.get(entity.id)?.get(component);
        if (currentValue !== baselineValue) {
          changed.push(component);
        }
      }
      return changed;
    },
  };
}

File: runtime-bevy/crates/threenative_runtime/src/component_diff.rs
use std::collections::{BTreeMap, BTreeSet};

use serde_json::Value;
use threenative_loader::{EntityComponents, LoadedBundle};

use crate::systems_context::{canonical_component_value, component_value};

#[derive(Debug, Default)]
pub struct ComponentDiffCache {
    baseline: BTreeMap<String, BTreeMap<String, String>>,
    tracked_components: BTreeSet<String>,
}

impl ComponentDiffCache {
    pub fn begin_schedule_stage(&mut self, bundle: &LoadedBundle, components: &[String]) {
        self.tracked_components = components.iter().cloned().collect();
        self.baseline.clear();
        for entity in &bundle.world.entities {
            let mut row = BTreeMap::new();
            for component in &self.tracked_components {
                if let Some(value) = component_value(&entity.components, component) {
                    row.insert(component.clone(), canonical_component_value(&value));
                }
            }
            if !row.is_empty() {
                self.baseline.insert(entity.id.clone(), row);
            }
        }
    }

    pub fn runtime_changed_components(
        &self,
        entity_id: &str,
        components: &EntityComponents,
    ) -> Vec<String> {
        self.tracked_components
            .iter()
            .filter_map(|component| {
                let current = component_value(components, component)
                    .map(|value| canonical_component_value(&value));
                let previous = self
                    .baseline
                    .get(entity_id)
                    .and_then(|row| row.get(component));
                (current.as_deref() != previous.map(String::as_str)).then(|| component.clone())
            })
            .collect()
    }

    pub fn runtime_changed_map(&self, bundle: &LoadedBundle) -> BTreeMap<String, Vec<String>> {
        bundle
            .world
            .entities
            .iter()
            .filter_map(|entity| {
                let changed = self.runtime_changed_components(&entity.id, &entity.components);
                (!changed.is_empty()).then_some((entity.id.clone(), changed))
            })
            .collect()
    }
}

pub fn changed_components(
    bundle: &LoadedBundle,
    entity_id: &str,
    components: &EntityComponents,
    diff_cache: Option<&ComponentDiffCache>,
) -> Vec<String> {
    let explicit = read_changed(components.extra.get("__changed"), entity_id)
        .into_iter()
        .chain(read_changed(
            bundle.world.resources.get("__changed"),
            entity_id,
        ))
        .chain(read_changed(
            bundle.world.resources.get("Changed"),
            entity_id,
        ))
        .collect::<Vec<_>>();
    if !explicit.is_empty() {
        return explicit;
    }
    diff_cache
        .map(|cache| cache.runtime_changed_components(entity_id, components))
        .unwrap_or_default()
}

fn read_changed(value: Option<&Value>, entity_id: &str) -> Vec<String> {
    let Some(value) = value else {
        return Vec::new();
    };
    if let Some(items) = value.as_array() {
        return items
            .iter()
            .filter_map(|item| item.as_str().map(str::to_owned))
            .collect();
    }
    if let Some(items) = value.get(entity_id).and_then(Value::as_array) {
        return items
            .iter()
            .filter_map(|item| item.as_str().map(str::to_owned))
            .collect();
    }
    value
        .get("entities")
        .and_then(|entities| entities.get(entity_id))
        .and_then(Value::as_array)
        .map(|items| {
            items
                .iter()
                .filter_map(|item| item.as_str().map(str::to_owned))
                .collect()
        })
        .unwrap_or_default()
}
